pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,793 @@
+				{
+					future_scheduled_deletion_assets.push(asset.clone());
+					<ScheduledAssetDeletion<T>>::insert(
+						new_time_slot,
+						future_scheduled_deletion_assets,
+					);
+				} else {
+					let new_asset_list = vec![asset.clone()];
+					<ScheduledAssetDeletion<T>>::insert(new_time_slot, new_asset_list);
+				}
+			};
+		}
+
+		pub fn create_new_asset(
+			asset: Vec<u8>,
+			target_price: u128,
+			upper_bound: u16,
+			lower_bound: u8,
+			asset_owner: AccountOf<T>,
+			expiration_period: UnixTime,
+			number_of_assets: u8,
+		) -> Result<(), DispatchError> {
+			AssetTargetPrices::<T>::insert(asset.clone(), target_price);
+			let asset_metadatum = AssetMetadatum::<T> {
+				upper_bound,
+				lower_bound,
+				expiration_period,
+				asset_sudo: asset_owner.clone(),
+			};
+			AssetMetadata::<T>::insert(asset.clone(), asset_metadatum);
+			let new_time_slot = Self::get_current_time_slot()?.saturating_add(expiration_period);
+			<ScheduledAssetDeletion<T>>::insert(new_time_slot, vec![asset.clone()]);
+			AssetPrices::<T>::insert(asset.clone(), target_price);
+			let new_number_of_assets = number_of_assets + 1;
+			NumberOfAssets::<T>::put(new_number_of_assets);
+			Self::deposit_event(Event::AssetCreated { asset });
+			Ok(())
+		}
+
+		pub fn get_current_time_slot() -> Result<UnixTime, Error<T>> {
+			let now = <timestamp::Pallet<T>>::get().saturated_into::<UnixTime>();
+			if now == 0 {
+				Err(Error::<T>::BlockTimeNotSet)?
+			}
+			let now = now.saturating_div(1000);
+			let diff_to_min = now % 60;
+			Ok(now.saturating_sub(diff_to_min))
+		}
+
+		pub fn delete_asset_tasks(asset: Vec<u8>) {
+			// delete scheduled tasks
+			ScheduledTasks::<T>::remove_prefix((asset.clone(),), None);
+			// delete tasks from tasks table
+			Tasks::<T>::remove_prefix((asset.clone(),), None);
+			// delete tasks from task queue
+			let existing_task_queue: Vec<(Vec<u8>, T::Hash)> = Self::get_task_queue();
+			let mut updated_task_queue: Vec<(Vec<u8>, T::Hash)> = vec![];
+			for task in existing_task_queue {
+				if task.0 != asset {
+					updated_task_queue.push(task);
+				}
+			}
+			TaskQueue::<T>::put(updated_task_queue);
+		}
+
+		pub fn run_notify_task(message: Vec<u8>) -> Weight {
+			Self::deposit_event(Event::Notify { message });
+			<T as Config>::WeightInfo::emit_event()
+		}
+
+		/// Runs as many tasks as the weight allows from the provided vec of task_ids.
+		///
+		/// Returns a vec with the tasks that were not run and the remaining weight.
+		pub fn run_tasks(
+			mut task_ids: Vec<(Vec<u8>, T::Hash)>,
+			mut weight_left: Weight,
+		) -> (Vec<(Vec<u8>, T::Hash)>, Weight) {
+			let mut consumed_task_index: usize = 0;
+			for task_id in task_ids.iter() {
+				consumed_task_index.saturating_inc();
+				let action_weight = match Self::get_task(task_id) {
+					None => {
+						Self::deposit_event(Event::TaskNotFound { task_id: task_id.1.clone() });
+						<T as Config>::WeightInfo::emit_event()
```

Comment:
is this also a read?

Reply:
it seems like yes. e.g. https://github.com/paritytech/substrate/blob/master/bin/node-template/pallets/template/src/lib.rs#L84. there's a read no matter what, even if not found